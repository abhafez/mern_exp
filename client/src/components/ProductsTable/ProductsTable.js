import React from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';
import { Dropdown, Segment, Table } from 'semantic-ui-react';
import Pagination from 'components/Pagination';
import { options } from 'constants/PaginationOptions';

function ProductsTable(props) {
  const {
    tableHeaders,
    productList,
    itemsPerPage,
    handlItemsPerPage,
    maxLength,
    pageNumber,
    setPageNumber,
  } = props;

  return (
    <Table celled textAlign="center">
      <Table.Header>
        <Table.Row>
          {tableHeaders.map((title) => (
            <Table.HeaderCell key={title}>
              <FormattedMessage id={`product.${title}`} />
            </Table.HeaderCell>
          ))}
        </Table.Row>
      </Table.Header>

      <Table.Body>
        {productList.map((product) => (
          <Table.Row key={product.id}>
            <Table.Cell>{product.name}</Table.Cell>
            <Table.Cell>{product.newPrice ? product.newPrice : product.price}</Table.Cell>
            <Table.Cell>{product.code}</Table.Cell>
            <Table.Cell>{product.department}</Table.Cell>
          </Table.Row>
        ))}
      </Table.Body>

      <Table.Footer>
        <Table.Row>
          <Table.HeaderCell colSpan="4">
            <Segment basic floated="left">
              <Dropdown
                options={options}
                selection
                value={itemsPerPage}
                onChange={handlItemsPerPage}
              />
            </Segment>
            <Segment basic floated="right">
              <Pagination
                numberOfItems={maxLength}
                itemsPerPage={itemsPerPage}
                pageNumber={pageNumber}
                onChangePage={(pageNum) => {
                  setPageNumber(pageNum);
                }}
              />
            </Segment>
          </Table.HeaderCell>
        </Table.Row>
      </Table.Footer>
    </Table>
  );
}

ProductsTable.propTypes = {
  tableHeaders: PropTypes.object,
  productList: PropTypes.object,
  itemsPerPage: PropTypes.number,
  handlItemsPerPage: PropTypes.func,
  maxLength: PropTypes.number,
  pageNumber: PropTypes.number,
  setPageNumber: PropTypes.func,
};

export default ProductsTable;
