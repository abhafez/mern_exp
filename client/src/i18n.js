/** i18n setup */

/**
 * Getting messages
 */
const enTranslationMessages = require('translations/en.json');

/**
 * Getting and saving default locale for the project
 * You can edit in constants file only
 */
const { DEFAULT_LOCALE } = require('constants/locale');
localStorage.setItem('locale', DEFAULT_LOCALE);

const formatTranslationMessages = (locale, messages) => {
  const defaultFormattedMessages =
    locale !== DEFAULT_LOCALE
      ? formatTranslationMessages(DEFAULT_LOCALE, enTranslationMessages)
      : {};
  const flattenFormattedMessages = (formattedMessages, key) => {
    const formattedMessage =
      !messages[key] && locale !== DEFAULT_LOCALE
        ? defaultFormattedMessages[key]
        : messages[key];
    return Object.assign(formattedMessages, { [key]: formattedMessage });
  };
  return Object.keys(messages).reduce(flattenFormattedMessages, {});
};

export const translationMessages = {
  en: formatTranslationMessages('en', enTranslationMessages),
};
