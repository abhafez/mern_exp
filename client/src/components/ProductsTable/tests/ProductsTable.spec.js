import { render, screen } from '@testing-library/react';
import { IntlProvider } from 'react-intl';
import ProductsTable from '../index';
import { translationMessages } from '../../../i18n';

describe('<ProductsTable />', () => {
  it('Should be able to update page number', () => {
    const tableHeaders = ['name', 'price', 'code', 'department'];
    const productList = [{ id: 321, name: 'PC', price: 30000 }];
    const handleItemsPerPageSpy = jest.fn();
    const setPageSpy = jest.fn();
    const EN = 'en';
    const { container } = render(
      <IntlProvider locale={EN} messages={translationMessages[EN]}>
        <ProductsTable
          tableHeaders={tableHeaders}
          productList={productList}
          itemsPerPage={10}
          maxLength={10}
          handlItemsPerPage={handleItemsPerPageSpy}
          setPageNumber={setPageSpy}
          pageNumber={1}
        />
      </IntlProvider>,
    );
    expect(handleItemsPerPageSpy).not.toHaveBeenCalled();
    expect(setPageSpy).toHaveBeenCalled(); // by the pagination component

    expect(container.firstChild).toMatchSnapshot();

    // Pagination
    expect(screen.getByText('1').classList).toContain('active');
    expect(screen.getByText('⟩⟩').classList).toContain('disabled');
  });
});
