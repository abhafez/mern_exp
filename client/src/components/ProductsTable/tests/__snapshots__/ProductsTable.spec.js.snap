// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductsTable /> Should be able to update page number 1`] = `
<table
  class="ui celled center aligned table"
>
  <thead
    class=""
  >
    <tr
      class=""
    >
      <th
        class=""
      >
        Product Name
      </th>
      <th
        class=""
      >
        Product Price
      </th>
      <th
        class=""
      >
        Promotion Code
      </th>
      <th
        class=""
      >
        Department
      </th>
    </tr>
  </thead>
  <tbody
    class=""
  >
    <tr
      class=""
    >
      <td
        class=""
      >
        PC
      </td>
      <td
        class=""
      >
        30000
      </td>
      <td
        class=""
      />
      <td
        class=""
      />
    </tr>
  </tbody>
  <tfoot
    class=""
  >
    <tr
      class=""
    >
      <th
        class=""
        colspan="4"
      >
        <div
          class="ui basic left floated segment"
        >
          <div
            aria-expanded="false"
            class="ui selection dropdown"
            role="listbox"
            tabindex="0"
          >
            <div
              aria-atomic="true"
              aria-live="polite"
              class="text"
              role="alert"
            />
            <i
              aria-hidden="true"
              class="dropdown icon"
            />
            <div
              class="menu transition"
            >
              <div
                aria-checked="false"
                aria-selected="true"
                class="selected item"
                role="option"
                style="pointer-events: all;"
              >
                <span
                  class="text"
                >
                  5 Items
                </span>
              </div>
              <div
                aria-checked="false"
                aria-selected="false"
                class="item"
                role="option"
                style="pointer-events: all;"
              >
                <span
                  class="text"
                >
                  8 Items
                </span>
              </div>
              <div
                aria-checked="false"
                aria-selected="false"
                class="item"
                role="option"
                style="pointer-events: all;"
              >
                <span
                  class="text"
                >
                  12 Items
                </span>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ui basic right floated segment"
        >
          <div
            aria-label="Pagination Navigation"
            class="ui pagination menu"
            id="pagination-element"
            role="navigation"
          >
            <div
              aria-current="false"
              aria-disabled="false"
              aria-label="Previous item"
              class="disabled item"
              id="first-page"
              role="button"
              tabindex="0"
              type="firstItem"
              value="1"
            >
              ⟨⟨
            </div>
            <div
              aria-current="false"
              aria-disabled="false"
              aria-label="Previous item"
              class="disabled item"
              role="button"
              tabindex="0"
              type="prevItem"
            >
              ⟨
            </div>
            <div
              aria-current="true"
              aria-disabled="false"
              class="active item"
              role="button"
              tabindex="0"
              type="pageItem"
              value="1"
            >
              1
            </div>
            <div
              aria-current="false"
              aria-disabled="false"
              aria-label="Next item"
              class="disabled item"
              role="button"
              type="nextItem"
            >
              ⟩
            </div>
            <div
              aria-current="false"
              aria-disabled="false"
              aria-label="Next item"
              class="disabled item"
              tabindex="0"
              type="nextItem"
            >
              ⟩⟩
            </div>
          </div>
        </div>
      </th>
    </tr>
  </tfoot>
</table>
`;
